server/src/main/java/io/druid/client/CachingClusteredClient.java

```diff
@@ -221,369 +300,329 @@ public CachingClusteredClient(
-                                                          return cacheFn.apply(input);
-                                                        }
-                                                      }
-                                                  )
-                                              );
-                                            }
-                                            return input;
-                                          }
-                                        }
-                                    ),
-                                    toolChest.makePreComputeManipulatorFn(
-                                        // Ick... most makePreComputeManipulatorFn directly cast to their ToolChest query type of choice
-                                        // This casting is sub-optimal, but hasn't caused any major problems yet...
-                                        (Query) rewrittenQuery,
-                                        MetricManipulatorFns.deserializing()
-                                    )
-                                ),
-                                new Runnable()
-                                {
-                                  @Override
-                                  public void run()
-                                  {
-                                    if (cachePopulator != null) {
-                                      Futures.addCallback(
-                                          Futures.allAsList(cacheFutures),
-                                          new FutureCallback<List<Object>>()
-                                          {
-                                            @Override
-                                            public void onSuccess(List<Object> cacheData)
-                                            {
-                                              cachePopulator.populate(cacheData);
-                                              // Help out GC by making sure all references are gone
-                                              cacheFutures.clear();
-                                            }
-
-                                            @Override
-                                            public void onFailure(Throwable throwable)
-                                            {
-                                              log.error(throwable, "Background caching failed");
-                                            }
-                                          },
-                                          backgroundExecutorService
-                                      );
-                                    }
-                                  }
-                                },
-                                MoreExecutors.sameThreadExecutor()
-                            );// End withEffect
-                          }
-                        }
-                    )
+    private Sequence<T> getAndCacheServerResults(
```

Comment:
It doesn't actually "gets bySegment results", there is another method "getBySegmentServerResults" for this. This method "gets non-by segment results", and it's "implementation detail" that it converts the query to by-segment and caches the results by-segment.

Naming of methods "getBySegmentServerResults", "getSimpleServerResults" and "getAndCacheServerResults" is not perfect but I cannot find a better set of names for them